lib/Sema/TypeCheckDecl.cpp

```diff
@@ -4727,36 +4721,37 @@ void TypeChecker::validateDeclForNameLookup(ValueDecl *D) {
       if (!typealias->getGenericParams()) {
         if (typealias->isBeingValidated()) return;
 
-        bool validated = typealias->hasValidationStarted();
+        auto helper = [&] {
+          (void) typealias->getFormalAccess();
 
-        if (!validated)
-          typealias->setIsBeingValidated();
-        SWIFT_DEFER {
-          if (!validated)
-            typealias->setIsBeingValidated(false);
-        };
+          ProtocolRequirementTypeResolver resolver;
+          TypeResolutionOptions options =
+            TypeResolutionFlags::TypeAliasUnderlyingType;
+          if (validateType(typealias->getUnderlyingTypeLoc(),
+                           typealias, options, &resolver)) {
+            typealias->setInvalid();
+            typealias->getUnderlyingTypeLoc().setInvalidType(Context);
+          }
 
-        (void) typealias->getFormalAccess();
+          typealias->setUnderlyingType(
+                                  typealias->getUnderlyingTypeLoc().getType());
 
-        ProtocolRequirementTypeResolver resolver;
-        TypeResolutionOptions options =
-          TypeResolutionFlags::TypeAliasUnderlyingType;
-        if (validateType(typealias->getUnderlyingTypeLoc(),
-                         typealias, options, &resolver)) {
-          typealias->setInvalid();
-          typealias->getUnderlyingTypeLoc().setInvalidType(Context);
-        }
+          // Note that this doesn't set the generic environment of the alias yet,
+          // because we haven't built one for the protocol.
+          //
+          // See how validateDecl() sets the generic environment on alias members
+          // explicitly.
+          //
+          // FIXME: Hopefully this can all go away with the ITC.
```

Comment:
The ITC is now gone. Our hopes and dreams are pinned on the request-evaluator ;)